doc/manual/src/store/file-system-object/content-address.md

```diff
@@ -0,0 +1,83 @@
+The various *methods* of content addressing thus differ in how abstract data (in this case, a file system object and its descendents) are fed into the hash function.
+
+## Serialising File System Objects { #serial }
+
+The simplest method is to serialise the entire file system object tree into a single binary string, and then hash that binary string, yielding the content address.
+In this section we describe the extent methods of serialising file system objects.
+
+### Flat { #serial-flat }
+
+A single file object can just be hashed by its contents.
+This is not enough information to encode the fact that the file system object is a file,
+but if we *already* know that the FSO is a single non-executable file by other means, it is sufficient.
+
+### Nix Archive (NAR) { #serial-nix-archive }
+
+For the other cases of FSOs, especially directories with arbitrary descendents, we need a more complex serialisation format.
+Examples of such serialisations are the ZIP and TAR file formats.
+However, for our purposes these formats have two problems:
+
+- They do not have a canonical serialisation, meaning that given an FSO, there can
+be many different serialisations.
+  For instance, TAR files can have variable amounts of padding between archive members;
+  and some archive formats leave the order of directory entries undefined.
+  This is bad because we use serialisation to compute cryptographic hashes over FSOs, and therefore require the serialisation to be unique.
+  Otherwise, the hash value can depend on implementation details or environment
+settings of the serialiser.
+
+- They store more information than we have in our notion of FSOs, such as time stamps.
+  This can cause FSOs that Nix should consider equal to hash to different values on different machines, just because the dates differ.
+
+- As a practical consideration, the TAR format is the only truly universal format in the Unix environment.
+  It has many problems, such as an inability to deal with long file names and files larger than 2^33 bytes.
+  Current implementations such as GNU Tar work around these limitations in various ways.
+
+For these reasons, Nix has its very own archive format—the Nix Archive (NAR) format,
+which is carefully designed to avoided the problems described above.
+
+The exact specification of the Nix Archive format is in `protocols/nix-archive.md`
+
+## Content addressing File System Objects beyond a single serialisation pass
+
+Serialising the entire tree and then hashing that binary string is not the only option for content addressing, however.
+There is also the [Merkle DAG](https://en.wikipedia.org/wiki/Merkle_tree) of multiple rounds of serialisation and hashing, where previously computed hashes are included in subsequent byte strings to be hashed.
+
+This creates the possibility of hashing reflecing the graph structure of file system objects:
```

Comment:
I'm a bit confused, I did *not* want to talk about structure in the previous paragraph because I want to talk about it in this paragraph --- I like the separation of concerns so the reader doesn't need to ingest too many concepts at once.

For me the conceptual order is:

1. We don't have to hash everything in one go
2. We can hash little by little matching the semantic graph structure in particular
3. Extra benefits from the above